ucp/_libs/core.pyx

```diff
@@ -403,10 +404,15 @@ class _Endpoint:
         cdef CtrlMsg msg
         cdef CtrlMsgData[::1] ctrl_msg_mv
         try:
+            # Making sure we only tell peer to shutdown once
+            if self._shutting_down_peer:
+                return
+            self._shutting_down_peer = True
```

Comment:
Maybe it would make more sense to move this block to the top of the function:

```python
if self._closed or self._shutting_down_peer:
    return
self._shutting_down_peer = True
```

Reply:
It is there to make sure we run:
```python
    finally:
            self._ctx.worker.progress()
            await asyncio.sleep(0)
            self.abort()
```

Reply:
Interesting, it seems logical now but I didn't realize `finally` would get executed even if there's a return inside the `try` block. Thanks for clarifying @madsbk !